File: src/Soel/Stages/Repair.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}

module Soel.Stages.Repair
  ( repairLoop
  ) where

import Control.Monad.Except (throwError)
import Control.Monad.IO.Class (liftIO)
import Control.Monad.Reader (ask)
import Data.Aeson ((.:), (.:?), withObject, eitherDecode)
import qualified Data.Aeson as Aeson
import Data.Aeson.Types (parseEither)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Lazy as TL
import qualified Data.Text.Lazy.Encoding as TLE
import System.Console.ANSI (SGR(..), Color(..), ColorIntensity(..), ConsoleLayer(..))
import System.Console.Haskeline (InputT, runInputT, defaultSettings, getInputLine)
import System.FilePath (replaceExtension)

import Soel.Types (App, SoelError(..), SoelConfig(..), runApp, requireApiKeyM)
import Soel.Config (OpenRouterConfig(..))
import Soel.IR.Types (SourceFile(..), CodeIR, AmbiguityResult(..))
import Soel.IR.Transform (transformToCodeIR)
import Soel.Stages.Reader (readSource)
import Soel.Stages.SemanticEncoder (semanticEncode)
import Soel.Stages.AmbiguityDetector (detectAmbiguities)
import Soel.Stages.Dialog (autoResolveAll)
import Soel.Stages.Codegen (generateHaskell, CodegenResult(..))
import Soel.Stages.Writer (writeHaskell, WriteOptions(..))
import Soel.Stages.GHC (ghcCompile, ghcRun, GHCResult(..))
import Soel.LLM.OpenRouter
  ( llmRequest, extractJSON
  , LLMRequestOptions(..), LLMMessage(..), LLMRole(..) )
import Soel.LLM.Prompts (getPrompt)
import Soel.Utils.Logger
  ( logStage, logInfo, logSuccess, logWarn, logError
  , spinnerSuccess, spinnerFail, coloredStderr )

--------------------------------------------------------------------------------
-- Repair result from LLM
--------------------------------------------------------------------------------

data RepairResult = RepairResult
  { rrDiagnosis      :: !Text
  , rrRootCause      :: !Text
  , rrFixType        :: !Text
  , rrFixedCode      :: !Text
  , rrExplanation    :: !Text
  , rrSoelSuggestion :: !(Maybe Text)
  } deriving (Show)

parseRepairResult :: Text -> Either String RepairResult
parseRepairResult txt =
  case eitherDecode (TLE.encodeUtf8 (TL.fromStrict txt)) of
    Left err  -> Left err
    Right val -> parseEither parser val
  where
    parser = withObject "RepairResult" $ \v -> do
      rrDiagnosis      <- v .:  "diagnosis"
      rrRootCause      <- v .:  "root_cause"
      rrFixType        <- v .:  "fix_type"
      rrFixedCode      <- v .:  "fixed_code"
      rrExplanation    <- v .:  "explanation"
      rrSoelSuggestion <- v .:? "soel_suggestion"
      pure RepairResult{..}

--------------------------------------------------------------------------------
-- Main entry point
--------------------------------------------------------------------------------

-- | Interactive repair loop: compiles a .soel file, then iteratively fixes
-- GHC errors by consulting the LLM. Up to 5 rounds.
repairLoop :: FilePath -> App ()
repairLoop filePath = do
  cfg <- ask
  liftIO $ logStage "Pragmatic repair mode"

  -- Step 1: Read source
  source <- readSource filePath
  let hsPath = replaceExtension (sfPath source) ".hs"

  liftIO $ coloredStderr [SetColor Foreground Dull White] "\n  Performing initial compilation..."

  -- Step 2: Full compile pipeline
  narrativeIR <- semanticEncode (sfContent source)
  codeIR      <- transformToCodeIR narrativeIR (sfContent source)
  ambResult   <- detectAmbiguities narrativeIR codeIR
  let resolved = autoResolveAll (arAll ambResult)
  CodegenResult{..} <- generateHaskell codeIR resolved (sfContent source)

  _ <- writeHaskell crHaskellSource WriteOptions
    { woSourcePath = sfPath source
    , woOutputPath = Just hsPath
    }

  -- Step 3: Enter interactive repair loop (runs in Haskeline for readline)
  liftIO $ runInputT defaultSettings $
    repairRounds cfg source codeIR hsPath crHaskellSource 1 maxRounds
  where
    maxRounds :: Int
    maxRounds = 5

--------------------------------------------------------------------------------
-- Repair rounds (runs inside Haskeline's InputT IO)
--------------------------------------------------------------------------------

repairRounds
  :: SoelConfig
  -> SourceFile
  -> CodeIR
  -> FilePath       -- ^ .hs output path
  -> Text           -- ^ current Haskell source
  -> Int            -- ^ current round (1-based)
  -> Int            -- ^ max rounds
  -> InputT IO ()
repairRounds cfg source codeIR hsPath haskellSrc rnd maxR
  | rnd > maxR = liftIO $ logWarn $
      "Max repair rounds (" <> T.pack (show maxR) <> ") reached"
  | otherwise = do
      liftIO $ coloredStderr [SetColor Foreground Vivid Cyan] $
        "\n\x2500\x2500\x2500 Repair round "
        <> T.pack (show rnd) <> "/" <> T.pack (show maxR)
        <> " \x2500\x2500\x2500"

      -- Try ghcCompile + ghcRun
      compileResult <- liftIO $ runApp cfg $ do
        result <- ghcCompile hsPath
        ghcRun (grExecutablePath result)

      case compileResult of
        Right () ->
          liftIO $ logSuccess "Program compiled and ran successfully!"

        Left (GHCError _msg ghcStderr _exitCode) -> do
          -- Show the GHC error
          liftIO $ do
            coloredStderr [SetColor Foreground Vivid Red] "\n  GHC Error:"
            coloredStderr [SetColor Foreground Dull White] $
              T.unlines $ map ("    " <>) (T.lines ghcStderr)

          -- Ask LLM to diagnose and fix
          repairResult <- liftIO $ runApp cfg $
            repairWithLLM source codeIR haskellSrc ghcStderr

          case repairResult of
            Left err -> do
              liftIO $ do
                spinnerFail "Repair analysis failed"
                logError $ T.pack (show err)
              repairRounds cfg source codeIR hsPath haskellSrc (rnd + 1) maxR

            Right repair -> do
              liftIO $ do
                spinnerSuccess "Fix generated"
                coloredStderr [SetColor Foreground Vivid Yellow] $
                  "\n  Diagnosis: " <> rrDiagnosis repair
                coloredStderr [SetColor Foreground Vivid White] $
                  "  Fix: " <> rrExplanation repair
                case rrSoelSuggestion repair of
                  Just sug -> coloredStderr [SetColor Foreground Dull White] $
                    "\n  SOEL suggestion: " <> sug
                  Nothing  -> pure ()

              -- Prompt user
              mAnswer <- getInputLine "\n  Apply fix? [Y/n] "
              let answer = T.strip $ T.pack $ maybe "" id mAnswer

              if T.toLower answer == "n"
                then do
                  liftIO $ logInfo "Fix skipped"
                  mCustom <- getInputLine
                    "  Describe the issue or press Enter to retry: "
                  case mCustom of
                    Just c | not (T.null (T.strip (T.pack c))) ->
                      liftIO $ coloredStderr [SetColor Foreground Dull White] $
                        "  Will incorporate: \"" <> T.pack c <> "\""
                    _ -> pure ()
                  repairRounds cfg source codeIR hsPath haskellSrc (rnd + 1) maxR
                else do
                  let newSrc = rrFixedCode repair
                  _ <- liftIO $ runApp cfg $
                    writeHaskell newSrc WriteOptions
                      { woSourcePath = sfPath source
                      , woOutputPath = Just hsPath
                      }
                  liftIO $ logInfo "Fix applied, retrying..."
                  repairRounds cfg source codeIR hsPath newSrc (rnd + 1) maxR

        Left otherErr -> do
          -- Non-GHC error
          liftIO $ logError $ "Unexpected error: " <> T.pack (show otherErr)

--------------------------------------------------------------------------------
-- LLM repair call
--------------------------------------------------------------------------------

repairWithLLM :: SourceFile -> CodeIR -> Text -> Text -> App RepairResult
repairWithLLM source codeIR haskellSrc ghcStderr = do
  cfg <- ask
  apiKey <- requireApiKeyM

  let systemPrompt = getPrompt "pragmatic-repair"
      userPayload  = Aeson.encode $ Aeson.object
        [ ("soel_source",  Aeson.String (sfContent source))
        , ("haskell_code", Aeson.String haskellSrc)
        , ("ghc_error",    Aeson.String ghcStderr)
        , ("code_ir",      Aeson.toJSON codeIR)
        ]

  raw <- llmRequest LLMRequestOptions
    { lroApiKey      = apiKey
    , lroModel       = orModel (cfgOpenRouter cfg)
    , lroMessages    =
        [ LLMMessage RoleSystem systemPrompt
        , LLMMessage RoleUser (TL.toStrict $ TLE.decodeUtf8 userPayload)
        ]
    , lroTemperature = 0.2
    , lroMaxTokens   = Nothing
    , lroJsonMode    = False
    }

  let jsonText = extractJSON raw
  case parseRepairResult jsonText of
    Left err -> throwError $ CodegenError $
      "Failed to parse repair response: " <> T.pack err
    Right repair -> pure repair

File: src/stages/repair.ts
import { readFileSync, writeFileSync } from 'node:fs';
import { createInterface } from 'node:readline';
import { resolve, basename } from 'node:path';
import chalk from 'chalk';
import type { SoelConfig } from '../config.js';
import { llmRequest, extractJSON } from '../llm/openrouter.js';
import { loadPrompt } from '../llm/prompts.js';
import { readSource } from './reader.js';
import { semanticEncode } from './semantic-encoder.js';
import { transformToCodeIR } from '../ir/transform.js';
import { detectAmbiguities } from './ambiguity-detector.js';
import { autoResolveAll } from './dialog.js';
import { generateHaskell } from './codegen.js';
import { writeHaskell } from './writer.js';
import { ghcCompile, ghcRun } from './ghc.js';
import { GHCError } from '../utils/errors.js';
import { log, spinner } from '../utils/logger.js';

interface RepairResult {
  diagnosis: string;
  root_cause: string;
  fix_type: string;
  fixed_code: string;
  explanation: string;
  soel_suggestion?: string;
}

export async function repairLoop(
  filePath: string,
  config: SoelConfig
): Promise<void> {
  log.stage('Pragmatic repair mode');

  const source = readSource(filePath);
  const hsPath = source.path.replace(/\.soel$/, '.hs');

  // First, do a full compile
  console.error(chalk.dim('\n  Performing initial compilation...\n'));

  const narrativeIR = await semanticEncode(source.content, config);
  const codeIR = await transformToCodeIR(narrativeIR, source.content, config);
  const ambResult = detectAmbiguities(narrativeIR, codeIR, config);
  const resolved = autoResolveAll(ambResult.all);
  let { haskellSource } = await generateHaskell(codeIR, resolved, source.content, config);

  writeHaskell(haskellSource, { sourcePath: source.path, outputPath: hsPath });

  const rl = createInterface({
    input: process.stdin,
    output: process.stderr,
  });
  const ask = (prompt: string): Promise<string> =>
    new Promise((resolve) => rl.question(prompt, resolve));

  const maxRounds = 5;
  let round = 0;

  while (round < maxRounds) {
    round++;
    console.error(chalk.cyan(`\n─── Repair round ${round}/${maxRounds} ───\n`));

    try {
      const result = await ghcCompile(hsPath, config);
      await ghcRun(result.executablePath);

      log.success('Program compiled and ran successfully!');
      rl.close();
      return;
    } catch (e) {
      if (!(e instanceof GHCError)) throw e;

      console.error(chalk.red('\n  GHC Error:'));
      console.error(chalk.dim(e.stderr.split('\n').map((l) => `    ${l}`).join('\n')));

      // Ask LLM to repair
      const spin = spinner('Analyzing error and generating fix...').start();

      try {
        const systemPrompt = loadPrompt('pragmatic-repair');
        const raw = await llmRequest({
          apiKey: config.openrouter.apiKey!,
          model: config.openrouter.model,
          messages: [
            { role: 'system', content: systemPrompt },
            {
              role: 'user',
              content: JSON.stringify({
                soel_source: source.content,
                haskell_code: haskellSource,
                ghc_error: e.stderr,
                code_ir: codeIR,
              }),
            },
          ],
          temperature: 0.2,
          maxTokens: 16384,
        });

        const json = extractJSON(raw);
        const repair = JSON.parse(json) as RepairResult;

        spin.succeed('Fix generated');

        console.error(chalk.yellow(`\n  Diagnosis: ${repair.diagnosis}`));
        console.error(chalk.white(`  Fix: ${repair.explanation}`));

        if (repair.soel_suggestion) {
          console.error(chalk.dim(`\n  SOEL suggestion: ${repair.soel_suggestion}`));
        }

        const answer = await ask(chalk.green('\n  Apply fix? [Y/n] '));

        if (answer.trim().toLowerCase() !== 'n') {
          haskellSource = repair.fixed_code;
          writeHaskell(haskellSource, { sourcePath: source.path, outputPath: hsPath });
          log.info('Fix applied, retrying...');
        } else {
          log.info('Fix skipped');
          const custom = await ask(chalk.green('  Describe the issue or press Enter to retry: '));
          if (custom.trim()) {
            // Send custom feedback to LLM in next round
            console.error(chalk.dim(`  Will incorporate: "${custom}"`));
          }
        }
      } catch (repairErr) {
        spin.fail('Repair analysis failed');
        console.error(chalk.red(`  ${repairErr}`));
      }
    }
  }

  log.warn(`Max repair rounds (${maxRounds}) reached`);
  rl.close();
}
